File: day12/solution.py
from collections import defaultdict

def parseGraph(lines):
    graph = defaultdict(lambda: set())
    for line in lines:
        endpoints = line.split('-')
        assert len(endpoints) == 2
        graph[endpoints[0]].add(endpoints[1].strip())
        graph[endpoints[1].strip()].add(endpoints[0])
    return graph

def traverseNoSecondSmall(graph, path=['start'], smalls=set(), debug=False):
    ''' traverse(graph, prefix): returns all paths that start with prefix and end at 'end' '''
    if debug:
        print(f'traversing from {",".join(path)}')
    if path[-1] == 'end':
        return [path]
    completePaths = []
    for outEdge in graph[path[-1]] - {'start'}:
        if outEdge in smalls:
            continue
        newSmalls = smalls.union({outEdge}) if outEdge.islower() else smalls
        completePaths = completePaths + traverseNoSecondSmall(
            graph, path + [outEdge], newSmalls, debug)
    return completePaths

def part1(filename):
    f = open(filename, 'r')
    lines = f.readlines()
    f.close()
    return len(traverseNoSecondSmall(parseGraph(lines)))

assert part1('sample.txt') == 10
assert part1('full.txt') == 4659

def traverseOneSecondSmall(graph, path=['start'], smalls=set(), secondSmall=None, debug=False):
    ''' traverse(graph, prefix): returns all paths that start with prefix and end at 'end' '''
    if debug:
        print(f'{",".join(path)}. Can reach {" ".join(graph[path[-1]])}')
    if path[-1] == 'end':
        return [path]
    completePaths = []
    for outEdge in graph[path[-1]] - {'start'}:
        newSecondSmall = secondSmall
        if outEdge in smalls:
            if secondSmall:
                if debug:
                    print(f'skipping {outEdge}, already double-visited {secondSmall}')
                continue
            newSecondSmall = outEdge
            if debug:
                print(f'allowed double-visit of {outEdge}')
        newSmalls = smalls.union({outEdge}) if outEdge.islower() else smalls
        completePaths = completePaths + traverseOneSecondSmall(
            graph, path + [outEdge], newSmalls, newSecondSmall, debug)
    return completePaths

def part2(filename):
    f = open(filename, 'r')
    lines = f.readlines()
    f.close()
    return len(traverseOneSecondSmall(parseGraph(lines)))

assert part2('sample.txt') == 36
assert part2('full.txt') == 148962

File: day12/solution.hs
import qualified Data.MultiMap as MM
import Data.List
import Data.Maybe
import Debug.Trace
import Split
import Text.Printf

type Graph = MM.MultiMap String String
type Node = String
type Path = [Node]

parseGraph :: [String] -> Graph
parseGraph [] = MM.empty
parseGraph (line:lines) = MM.insert from to $ MM.insert to from $ parseGraph lines
    where [from, to] = split "-" line

maybeAddSmall :: Node -> [Node] -> [Node]
maybeAddSmall cave smalls
    | all (\c -> elem c ['a'..'z']) cave = cave : smalls
    | otherwise = smalls

outEdges :: Graph -> Node -> [Node]
outEdges graph v = [oe | oe <- MM.lookup v graph, oe /= "start"]

-- traverseNoSecondSmall graph pathPrefix smalls -> [path]
traverseNoSecondSmall :: Graph -> Path -> [Node] -> [Path]
traverseNoSecondSmall _ path@("end":_) _ = [reverse path]
traverseNoSecondSmall graph (h:t) smalls
    | elem h smalls = []
    | otherwise     = foldr (++) [] outPaths
    where newSmalls = maybeAddSmall h smalls
          outPaths = [traverseNoSecondSmall graph (oe : h : t) newSmalls | oe <- outEdges graph h]

part1 :: String -> Int
part1 input = length $ traverseNoSecondSmall graph ["start"] []
    where graph = parseGraph $ split "\n" input

-- traverseOneSecondSmall graph pathPrefix smalls secondSmall -> [path]
traverseOneSecondSmall :: Graph -> Path -> [Node] -> Maybe Node -> [Path]
traverseOneSecondSmall _ path@("end":_) _ _ = [reverse path]
traverseOneSecondSmall graph (h:t) smalls ss@(Just _)
    | elem h smalls = []
    | otherwise     = foldr (++) [] outPaths
    where newSmalls = maybeAddSmall h smalls
          outPaths = [
            traverseOneSecondSmall graph (oe : h : t) newSmalls ss | oe <- outEdges graph h]
traverseOneSecondSmall graph (h:t) smalls Nothing
    | elem h smalls = foldr (++) [] [
            traverseOneSecondSmall graph (oe : h : t) newSmalls (Just h) | oe <- outEdges graph h]
    | otherwise     = foldr (++) [] [
            traverseOneSecondSmall graph (oe : h : t) newSmalls Nothing | oe <- outEdges graph h]
    where newSmalls = maybeAddSmall h smalls

part2 :: String -> Int
part2 input = length $ traverseOneSecondSmall graph ["start"] [] Nothing
    where graph = parseGraph $ split "\n" input

main = do
    sample <- readFile "sample.txt"
    full <- readFile "full.txt"
    putStrLn $ "part1(sample): " ++ (show $ part1 sample)
    putStrLn $ "part1(full): " ++ (show $ part1 full)
    putStrLn $ "part2(sample): " ++ (show $ part2 sample)
    putStrLn $ "part2(full): " ++ (show $ part2 full)
